[김은혜] Step3 로또(자동) PR 요청

eunveloper opened the pull request:
안녕하세요! 오늘 우테캠 첫번째 라이브 강의를 듣고 구현을 해보니
막히던 부분이 약간 풀리는 느낌이 들어서 정말 뿌듯했습니다 >_<!

한가지 질문이 있는데, 소스코드에 하드코딩적인 부분을 걷어내려고 클래스에 final 변수를 사용하곤 하는데,
LottoResult 클래스에 메세지를 생성하는 과정에서 저런 단발성 데이터들도 전부 final 변수로 사용하는것이 좋을까요?
특히 ( 이런 한가지 특수문자만 있는 경우에는 변수명 짓기도 꽤나 골치 아프더라구요...

좋은 아이디어나 생각 있으시면 리뷰 부탁드립니다! 감사합니다~ :)

eunveloper committed:
refactor: step2 피드백 반영

- String[] -> List<String> 으로 변경

eunveloper committed:
docs: README.md 작성

- step3

eunveloper committed:
docs: README.md 작성

- step3 과제에 대한 각종 요건사항 작성
- 로또 자동 프로그래밍 과제에 대한 구현할 기능목록 작성

eunveloper committed:
feat: 입력받은 구매금액으로 로또 생성하는 기능

- 테스트코드 작성
- 작성된 테스트 코드 기반으로 비지니스 로직 개발
- 비지니스 로직에 필요한 부가 로직 개발

eunveloper committed:
feat: 입력받은 구매금액으로 로또 생성하는 기능

- 비지니스 로직 변경으로 인한 테스트 코드 변경
- 필요한 View 클래스 및 기능 구현
- 도메인 클래스별로 구현해야할 행위를 정확히 나눠 기능 재구현

eunveloper committed:
feat: 입력받은 지난 주 당첨 번호로 통계를 내는 기능

- 필요한 테스트코드 작성
- 테스트코드 기반한 비지니스 로직 작성
- 테스트 코드를 위한 로직은 생성자를 활용
- 객체간의 비교를 위한 equals, hashcode 오버라이딩 추가
- 객체 필드에 직접 접근하지 않고 메세지를 활용

eunveloper committed:
feat: 입력받은 지난 주 당첨 번호로 통계를 내는 기능

- 기능 관련한 View 기능 작성
- 비교 결과 담는 객체 생성 및 기능 작성
- 비교 통계 기능 작성

eunveloper committed:
refactor: 하드코딩적인 부분 걷어내기

wheejuni commented on src/main/java/step3/constant/WinningPrice.java:
```diff
@@ -0,0 +1,17 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class WinningPrice {
```
`class` 보다 `enum` 이 더 좋진 않을까요?

eunveloper replied:
Class로 사용할지 Enum을 사용할지 갈등했었는데, 
Map 자료구조를 통해서 데이터를 한번에 조회하려는 생각으로 접근하다보니 클래스로 구성이 되었어요!
Enum으로 변경작업 해볼게요 👏

wheejuni commented on src/main/java/step3/constant/WinningPrice.java:
```diff
@@ -0,0 +1,17 @@
+
+public class WinningPrice {
+
+    public static final Map<Integer, Integer> price = new HashMap<>();
```
`enum` 을 사용한다면 여기도 걷어낼 수 있을 것 같긴 해요.

wheejuni commented on src/main/java/step3/model/Lotto.java:
```diff
@@ -0,0 +1,88 @@
+
+public class Lotto {
+
+    private final static int ZERO = 0;
```
`ZERO` 와 `0` 이 의미상으로 어떤 추상화를 거친 건지 잘 모르겠긴 합니다.

eunveloper replied:
다른 숫자부분들과 같이 하드코딩적인 부분을 걷어내려다 보니
뭔가 의미없는 추상화가 된거같네요 ㅠㅠ 고민해보겠습니다!

wheejuni commented on src/main/java/step3/model/Lotto.java:
```diff
@@ -0,0 +1,88 @@
+
+    private List<LottoNumber> numbers;
+
+    public Lotto() {
+        List<LottoNumber> allNumbers = settingLottoRangeNumber();
+        Collections.shuffle(allNumbers);
+
+        this.numbers = allNumbers.subList(ZERO, LOTTO_SIZE);
+        Collections.sort(this.numbers);
+    }
```
`settingLottoRangeNumber()` 는 약간의 시간 복잡도를 갖는 로직인데 생성자에 이게 위치해야 하는건지는 잘 모르겠습니다.
생성자는 객체 값의 할당에만 집중하면 어떨까요? 생성자에서 너무 많은 일이 일어나네요.
이 로직을 별도의 컴포넌트로 분리시키는 구현을 추천드려요.

eunveloper replied:
생성자는 객제 값의 할당에만 집중! 
좋은 내용 알아갑니다 ㅎㅎ 수정작업 해볼게요!

wheejuni commented on src/main/java/step3/model/Lotto.java:
```diff
@@ -0,0 +1,88 @@
+        Collections.sort(this.numbers);
+    }
+
+    public Lotto(String inputNumber) {
```
마찬가지 의견 드려요.

wheejuni commented on src/main/java/step3/model/Lotto.java:
```diff
@@ -0,0 +1,88 @@
+        return result;
+    }
+
+    private List<LottoNumber> settingLottoRangeNumber() {
+        List<LottoNumber> allNumbers = new ArrayList<>();
+        for (int i = LOTTO_START; i <= LOTTO_END; i++) {
+            allNumbers.add(new LottoNumber(i));
+        }
+        return allNumbers;
+    }
```
취향 문제이긴 하지만 학습 차원에서 `IntStream` 사용으로 개선해 보시는 건 어떨까요?

eunveloper replied:
Stream 문법이 익숙치 않네요 ㅠㅠ 이번 우테캠 기회로 연습해보겠습니다!

wheejuni commented on src/main/java/step3/model/Lotto.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    public int sameNumberCount(Lotto lotto) {
+        int sameCount = ZERO;
+        for (int i = ZERO; i < lotto.numbers.size(); i++) {
+            sameCount += containNumber(lotto.numbers.get(i));
+        }
+        return sameCount;
```
Stream 문법으로 개선해보시면 어떨까 제안드립니다.
`anyMatch()`, `count()` 등을 키워드로 검색하셔서 개선해 보세요.

wheejuni commented on src/main/java/step3/model/LottoResult.java:
```diff
@@ -0,0 +1,57 @@
+
+    public String createSameCountMessage(int sameNumber) {
+        return new StringBuilder(String.valueOf(sameNumber))
+                .append("개 일치 ")
+                .append("(")
+                .append(winningPrice(sameNumber))
+                .append("원")
+                .append(")- ")
+                .append(Optional.ofNullable(sameCounts.get(sameNumber)).orElse(0))
+                .append("개").toString();
```
굳이 이렇게 `.append()` 로 체이닝 하실 필요 없습니다. 
`+` 연산자로 묶으셔도 되고요, 이것이 어떻게 JVM 내부에서 최적화 되는지도 찾아보시면 좋겠네요.

eunveloper replied:
문자열 합치기를 할때 불변객체 메모리 이슈로 늘 append를 사용하는 버릇이 있었는데,
살짝 찾아보니 String constant pool 이라는 것이 있네요! 한번 제대로 공부 해봐야 겠어요.

휘준님께서는 어떤 상황에사 + 연산자를 어떤 상황에서 append() 를 사용하는게 좋은 상황이라고 생각하시나요?ㅠㅠ
적재적소에 기술을 사용하는게 참 어려운거 같습니다....

wheejuni replied:
@eunveloper for loop로 순회해야 하는 경우를 빼곤 `StringBulder` 를 명시적으로 사용해야 하는 이유를 잘 모르겠기는 합니다만..
constant pool 이전에 문자열을 + 연산자로 이으면 컴파일러가 내부적으로 `StringBuilder` 를 사용하는 것처럼 최적화를 해 줍니다.

wheejuni reviewed (changes requested):
수고하셨습니다. 몇 가지 개선 제안 드립니다 👍

eunveloper committed:
refactor: 코드리뷰 피드백 반영

- WinningPrice 클래스 방식에서 Enum 타입으로 변경
- 생성자는 객체를 생성만하게 연산작업은 따로 컴포넌트 분리
- 문자열 합치기 방식 변경
- stream 활용
- 테스트코드 변경

eunveloper committed:
refactor: 코드리뷰 피드백 반영

- 미사용 메소드 제거

wheejuni commented on src/main/java/step3/model/Lotto.java:
```diff
@@ -1,27 +1,34 @@
+    private String inputNumber;
 
-    public Lotto() {
+    public Lotto() {}
```
혹시 빈 생성자는 어떤 목적일까요?

wheejuni reviewed (approved):
수고하셨습니다. 💯
